hail/python/hail/experimental/vcf_combiner/vds_combiner.py

```diff
@@ -0,0 +1,190 @@
+
+            handled_fields['LA'] = hl.range(0, alleles_len - hl.if_else(has_non_ref, 1, 0))
+            handled_fields['LGT'] = get_lgt(e, alleles_len, has_non_ref, row)
+            if 'AD' in e:
+                handled_fields['LAD'] = hl.if_else(has_non_ref, e.AD[:-1], e.AD)
+            if 'PGT' in e:
+                handled_fields['LPGT'] = e.PGT
+            if 'PL' in e:
+                handled_fields['LPL'] = hl.if_else(has_non_ref,
+                                                   hl.if_else(alleles_len > 2,
+                                                              e.PL[:-alleles_len],
+                                                              hl.missing(e.PL.dtype)),
+                                                   hl.if_else(alleles_len > 1,
+                                                              e.PL,
+                                                              hl.missing(e.PL.dtype)))
+                handled_fields['RGQ'] = hl.if_else(
+                    has_non_ref,
+                    hl.if_else(e.GT.is_haploid(),
+                               e.PL[alleles_len - 1],
+                               e.PL[hl.call(0, alleles_len - 1).unphased_diploid_gt_index()]),
+                    hl.missing(e.PL.dtype.element_type))
+
+            handled_fields['gvcf_info'] = (hl.case()
+                                           .when(hl.is_missing(row.info.END),
+                                                 hl.struct(**(
+                                                     parse_as_fields(
+                                                         row.info.select(*info_to_keep),
+                                                         has_non_ref)
+                                                 )))
+                                           .or_missing())
+
+            pass_through_fields = {k: v for k, v in e.items() if k not in handled_names}
+            return hl.struct(**handled_fields, **pass_through_fields)
+
+        f = hl.experimental.define_function(
+            lambda row: hl.rbind(
+                hl.len(row.alleles), '<NON_REF>' == row.alleles[-1],
+                lambda alleles_len, has_non_ref: hl.struct(
+                    locus=row.locus,
+                    alleles=hl.if_else(has_non_ref, row.alleles[:-1], row.alleles),
+                    rsid=row.rsid,
+                    __entries=row.__entries.map(
+                        lambda e: make_entry_struct(e, alleles_len, has_non_ref, row)))),
+            mt.row.dtype)
+        _transform_variant_function_map[mt.row.dtype] = f
+    transform_row = _transform_variant_function_map[mt.row.dtype]
+    return Table(TableMapRows(mt._tir, Apply(transform_row._name, transform_row._ret_type, TopLevelReference('row'))))
+
+
+def make_reference_table(mt) -> Table:
+    mt = mt.filter(hl.is_defined(mt.info.END))
+
+    def make_entry_struct(e, row):
+        reference_fields = {k: v for k, v in e.items() if k in ('DP', 'GQ', 'MIN_DP')}
```

Comment:
We really can't hard-code this schema. I think the two options are to force the user to pass the reference block fields in explicitly, or compute them automatically from the data. I think I prefer the second, because it's the easiest user experience.

Reply:
Ok. In the update, I pushed a method to get the used fields for reference block entries. For combining we can populate it from an already extant VariantDataset, or from the first gvcf in order to not scan every gvcf.